### PR TITLE
ENH: Add errors parameter to DataFrame.rename

### LINKED ISSUE #13473: ENH: add errors='raise' option to rename
Is there a reason that `.rename` doesn't raise if keys are supplied that don't exist? This caught me out a couple of times:

``` python
In [14]: import pandas as pd

In [15]: df=pd.DataFrame({'a': [1,2], 'b': [3,4]})

In [16]: df
Out[16]: 
   a  b
0  1  3
1  2  4

In [17]: df.rename(columns={'a': 'c'})
Out[17]: 
   c  b
0  1  3
1  2  4
# ok

In [18]: df.rename(columns={'d': 'c'})
Out[18]: 
   a  b
0  1  3
1  2  4

# this should raise?
```

### PR DESCRIPTION
**Pull request by MaxVanDeursen**
- [X] closes #13473
- [x] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/frame/test_alter_axes.py [hunk 1]
```diff
@@ -871,6 +871,29 @@ def test_rename_bug2(self):
                              columns=["a"])
         tm.assert_frame_equal(df, expected)
 
+    def test_rename_errors(self):
+        # GH 13473
+        # rename now works with errors parameter
+
+        # Error has to be thrown and is thrown
+        df = DataFrame(columns=['A', 'B', 'C', 'D'])
```
Comment (WillAyd) [inline pandas/tests/frame/test_alter_axes.py]: We typically place these types of tests in a separate function with `_tests` appended to the end. Comment also not necessary

File: pandas/tests/frame/test_alter_axes.py [hunk 2]
```diff
@@ -871,6 +871,29 @@ def test_rename_bug2(self):
...
+        with pytest.raises(KeyError):
+            df.rename(columns={'A': 'a', 'E': 'e'}, errors='raise')
+
+        # Error should be ignored
+        renamed = df.rename(columns={'A': 'a', 'E': 'e'})
```
Comment (WillAyd) [inline pandas/tests/frame/test_alter_axes.py]: Can you parametrize these test cases

File: pandas/tests/frame/test_alter_axes.py [hunk 3]
```diff
@@ -871,6 +871,29 @@ def test_rename_bug2(self):
                              columns=["a"])
         tm.assert_frame_equal(df, expected)
 
+    def test_rename_errors(self):
+        # GH 13473
+        # rename now works with errors parameter
+
+        # Error has to be thrown and is thrown
+        df = DataFrame(columns=['A', 'B', 'C', 'D'])
+        with pytest.raises(KeyError):
```
Comment (WillAyd) [inline pandas/tests/frame/test_alter_axes.py]: Can you also add expected message here?

File: pandas/core/frame.py [hunk 4]
```diff
@@ -3924,30 +3925,45 @@ def rename(self, *args, **kwargs):
 
         Parameters
         ----------
-        mapper, index, columns : dict-like or function, optional
-            dict-like or functions transformations to apply to
+        mapper : dict-like or function
```
Comment (WillAyd) [inline pandas/core/frame.py]: OK as now I think but generally unrelated changes like this are better done in a separate PR

File: pandas/core/frame.py [hunk 5]
```diff
@@ -3924,30 +3925,45 @@ def rename(self, *args, **kwargs):
...
+
+        Raises
+        ------
+        KeyError
+            If any of the labels is not found in the selected axis.
 
         See Also
         --------
-        DataFrame.rename_axis
+        DataFrame.rename_axis: Set the name of the axis for the index or
```
Comment (WillAyd) [inline pandas/core/frame.py]: Minor but should require a space before the colon; also would be good if you can run the validation script against this docstring

File: pandas/core/generic.py [hunk 6]
```diff
@@ -981,11 +981,18 @@ def rename(self, *args, **kwargs):
...
+            If 'ignore', suppress error and existing labels are renamed.
 
         Returns
         -------
         renamed : %(klass)s (new object)
 
+        Raises
+        ------
+        KeyError
+            If any of the labels is not found in the selected axis.
```
Comment (WillAyd) [inline pandas/core/generic.py]: Should also mention `and ``errors='raise'```

**Review by WillAyd — CHANGES_REQUESTED** [review]
Nice thanks for the PR! Should add a whatsnew for 0.25 at this point

File: pandas/core/frame.py (hunk 4 above)
Comment (MaxVanDeursen) [reply pandas/core/frame.py]: I understand, I just figured it might be nice to reduce the technical debt of this method, as I saw that this docstring contained numerous errors. Will keep it in mind for next time!

File: pandas/core/frame.py (hunk 5 above)
Comment (MaxVanDeursen) [reply pandas/core/frame.py]: I ran the validation script against this docstring, which did not mark this as an error. However, I have adjusted this!

File: pandas/tests/frame/test_alter_axes.py (hunk 1 above)
Comment (MaxVanDeursen) [reply pandas/tests/frame/test_alter_axes.py]: Do you mean only the first added test (e.g. the test which checks whether an error is thrown) or all tests together? 

Moreover, I am unable to find a test which uses such naming pattern; would you mind giving me an example of one of these tests?

**Comment by MaxVanDeursen** [comment]
Would you want just one line in the Whatsnew enhancement section, or a whole subsection describing its use?

**Review by TomAugspurger — COMMENTED** [review]
A one-liner is fine for the enhancements section. That will link to the API docs.

Can you add an example using `errors` to the docstring?

File: pandas/tests/frame/test_alter_axes.py (hunk 3 above)
Comment (MaxVanDeursen) [reply pandas/tests/frame/test_alter_axes.py]: @TomAugspurger My bad, I normally do so when going through the feedback so that I know which ones I have fixed, after which I push my changes. Ofcourse, because I still had these questions, I now did not yet push the changes and hence they are marked resolved but not resolved in the repo.

Could you give me an example of the standard used syntax for test functions, as described by William? I couldn't find any tests which are appended with `_tests`

Comment (TomAugspurger) [reply pandas/tests/frame/test_alter_axes.py]: I'm not sure what @WillAyd was referring to in 
https://github.com/pandas-dev/pandas/pull/25535/files#r262267286. I'm guessing he'll clarify.

File: pandas/tests/frame/test_alter_axes.py (hunk 1 above)
Comment (WillAyd) [reply pandas/tests/frame/test_alter_axes.py]: Sorry I meant to say `_raises`. Here's one:

https://github.com/pandas-dev/pandas/blob/d10bbce349b293e9f753d799cd2ad032d5ec7333/pandas/tests/frame/test_alter_axes.py#L379

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/frame/test_alter_axes.py (hunk 1 above)
Comment (MaxVanDeursen) [reply pandas/tests/frame/test_alter_axes.py]: I see, that definitely clears things up! Thanks!

File: pandas/core/frame.py [hunk 7]
```diff
@@ -3924,30 +3925,46 @@ def rename(self, *args, **kwargs):
...
+        ------
+        KeyError
+            If any of the labels is not found in the selected axis and
+            "errors='raise'".
 
         See Also
         --------
-        DataFrame.rename_axis
+        DataFrame.rename_axis : Set the name of the axis for the index or
+         columns.
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: Check the doc build log for a warning. This may need to be one line. Maybe shorten to "Set the name of the axis"?

File: pandas/core/frame.py [hunk 8]
```diff
@@ -3924,30 +3925,46 @@ def rename(self, *args, **kwargs):
...
+        copy : bool, default True
+            Also copy underlying data.
+        inplace : bool, default False
             Whether to return a new DataFrame. If True then value of copy is
             ignored.
         level : int or level name, default None
             In case of a MultiIndex, only rename labels in the specified
             level.
+        errors : {'ignore', 'raise'}, default 'ignore'
+            If 'ignore', suppress error and existing labels are renamed.
```
Comment (TomAugspurger) [inline pandas/core/frame.py]: I think this could be clearer. Perhaps

    If 'raise', raise a `KeyError` when a dict-like `mapper`, `index`, or `columns` contains
    labels that are not present in the Index being transformed.

File: pandas/core/generic.py [hunk 9]
```diff
@@ -981,11 +981,19 @@ def rename(self, *args, **kwargs):
         level : int or level name, default None
             In case of a MultiIndex, only rename labels in the specified
             level.
+        errors : {'ignore', 'raise'}, default 'ignore'
+            If 'ignore', suppress error and existing labels are renamed.
```
Comment (TomAugspurger) [inline pandas/core/generic.py]: Same comment as above.

File: pandas/core/generic.py [hunk 10]
```diff
@@ -1085,10 +1094,19 @@ def rename(self, *args, **kwargs):
...
             f = com._get_rename_function(v)
-
             baxis = self._get_block_manager_axis(axis)
             if level is not None:
                 level = self.axes[axis]._get_level_number(level)
+
+            # GH 13473
+            if not callable(v):
+                indexer = self.axes[axis].get_indexer_for(v)
+                missing_labels = [label for index, label in enumerate(v)
```
Comment (TomAugspurger) [inline pandas/core/generic.py]: I'm a tad concerned about perf here. Can you instead check first if there are any missing labels with something like

```python
missing = indexer[indexer == -1]
if len(missing):
    ...
```

I'd like to avoid the iteration over `v`, which may be large.

but maybe measure the relative perf of the two approaches.

File: pandas/core/frame.py (hunk 8 above)
Comment (MaxVanDeursen) [reply pandas/core/frame.py]: Do you think it's best to include both the ignore and raise description?

Comment (TomAugspurger) [reply pandas/core/frame.py]: Yeah. Maybe append a "If 'ignore', extra keys will be ignored"? Something along those lines?

I just want to avoid the impression that passing "extra" keys is somehow an mis-use of `.rename`.

File: pandas/core/generic.py (hunk 10 above)
Comment (MaxVanDeursen) [reply pandas/core/generic.py]: I understand what you mean; I introduced `missing_labels` to generate a nice error message, where only the labels which are missing are in the error message.

I see two options:
- Remove `missing_labels` altogether and just place `v` inside the error message.
- Move `missing_labels` into the if-statement, so that the enumeration is only done in the case of an KeyError.

I'd personally prefer the latter, what do you think is best?

File: pandas/core/frame.py (hunk 7 above)
Comment (MaxVanDeursen) [reply pandas/core/frame.py]: Although I do not see an error or warning while building the documentation, I will shorten the description.

File: pandas/core/generic.py (hunk 10 above)
Comment (TomAugspurger) [reply pandas/core/generic.py]: Sorry if I wasn't clear. Including the missing labels in the error message is nice. Moving it under the `if` statement is best.

**Comment by MaxVanDeursen** [comment]
@WillAyd @TomAugspurger Any other feedback?

File: pandas/core/generic.py [hunk 11]
```diff
@@ -981,11 +981,23 @@ def rename(self, *args, **kwargs):
         level : int or level name, default None
             In case of a MultiIndex, only rename labels in the specified
             level.
+        errors : {'ignore', 'raise'}, default 'ignore'
```
Comment (WillAyd) [inline pandas/core/generic.py]: Not something that you should solve in this PR but we shouldn't need to duplicate docstrings here. Would be nice to use the Substitution decorator as required for this if you want to do a follow up PR

File: pandas/tests/frame/test_alter_axes.py [hunk 12]
```diff
@@ -871,6 +871,23 @@ def test_rename_bug2(self):
...
+
+    @pytest.mark.parametrize('mapper, errors, expected_columns', [
+        ({'A': 'a', 'E': 'e'}, 'ignore', ['a', 'B', 'C', 'D']),
+        ({'A': 'a'}, 'raise', ['a', 'B', 'C', 'D']),
+        (str.lower, 'raise', ['a', 'b', 'c', 'd'])])
+    def test_rename_errors(self, mapper, errors, expected_columns):
+        # GH 13473
+        # rename now works with errors parameter
+        df = DataFrame(columns=['A', 'B', 'C', 'D'])
+        renamed = df.rename(columns=mapper, errors=errors)
```
Comment (WillAyd) [inline pandas/tests/frame/test_alter_axes.py]: Very minor nit but let's change `renamed` to `result` to reflect pseudo-standard from other test cases

**Review by WillAyd — CHANGES_REQUESTED** [review]
Minor thing otherwise lgtm

File: pandas/core/generic.py (hunk 11 above)
Comment (MaxVanDeursen) [reply pandas/core/generic.py]: I'll do a followup PR on that! Would you mind giving me an example?

**Comment by MaxVanDeursen** [comment]
@WillAyd Ping; fixed the minor feedback. Would you mind giving me an example of the Subsitute decorator so that I can use it in my followup PR?

File: pandas/core/generic.py (hunk 11 above)
Comment (WillAyd) [reply pandas/core/generic.py]: Sure here's one:

https://github.com/pandas-dev/pandas/blob/221be3b4adde0f45927803b1c593b56d4678faeb/pandas/core/frame.py#L665

**Review by WillAyd — APPROVED** [review]
lgtm - @TomAugspurger off to you

**Review by TomAugspurger — APPROVED** [review]

**Comment by TomAugspurger** [comment]
Thanks @MaxVanDeursen!
